src/v/redpanda/application.cc

```diff
@@ -3113,6 +3114,24 @@ void application::start_runtime_services(
           .get();
     }
 
+    // Verify the enterprise license when trying to upgrade Redpanda.
+    // By this point during startup we have enough information to evaluate both
+    // the state of the license and what enterprise features are used.
+    // - If redpanda has been restarted on an existing node, we have already
+    //   loaded the feature table from the local snapshot in
+    //   application::load_feature_table_snapshot.
+    // - If this is a new node joining an existing cluster, by this point we
+    //   have received a controller snapshot from another node in the join
+    //   response and have replicated and replayed the the controller stm to the
+    //   last_applied offset received in the join_node_reply above.
```

Comment:
question: So we know that because we are at this point past these two spots:

https://github.com/redpanda-data/redpanda/blob/d1b08da9a865588ec97f145de0995fe20bb2d85f/src/v/redpanda/application.cc#L3098-L3104

https://github.com/redpanda-data/redpanda/blob/d1b08da9a865588ec97f145de0995fe20bb2d85f/src/v/redpanda/application.cc#L3110-L3118?

Reply:
Almost. The second code ref is correct, that corresponds to a fresh node starting the cluster.

However, for restarting nodes, the replay happens inside `controller::start`, which is also before this point during startup: https://github.com/redpanda-data/redpanda/blob/d1b08da9a865588ec97f145de0995fe20bb2d85f/src/v/cluster/controller.cc#L512-L560
https://github.com/redpanda-data/redpanda/blob/d1b08da9a865588ec97f145de0995fe20bb2d85f/src/v/redpanda/application.cc#L2946-L2954

I've added another half-sentence to point readers to `controller::start`.